usb: device_next: cdc_acm: Use a separate buffer pool for OUT

ndreys opened the pull request:
CDC ACM code will try to eagerly consume buffers from the buffer pool as long as there some data in it's TX FIFO. Since both IN and OUT transfers share the same buffer pool this make the code vulnerable to the "noisy neighbor" problem, where a influx of TX/IN traffic can completely deplete the buffer pool and cause a buffer allocation failure in cdc_acm_rx_fifo_handler(). Since there is no retry mechanism on buffer allocation failure there we end up stuck with USB controller NAK-ing every OUT transfer.

To avoid this problem, let's just split the buffer pools into IN/OUT. This way we know that an allocation failure in cdc_acm_rx_fifo_handler() implies that there are queued IN transfers which will re-arm the DMA in USB peripheral once they complete.

cc: @jfischer-no

ndreys committed:
usb: device_next: cdc_acm: Don't use UDC buffer for OUT transfers

Using shared buffer for OUT transfers makes the code vulnerable to
the "noisy neighbor" problem, where a high voulume of USB activity can
deplete the shared pool and cause a buffer allocation
failure in cdc_acm_rx_fifo_handler(). Since there is no retry
mechanism on buffer allocation failure there we end up stuck with USB
controller NAK-ing every OUT transfer.

To avoid this problem, modify the code to always allocate OUT
transfers from the per-instance pool and only use the shared pool for
IN transfers. This way we know that an allocation failure in
cdc_acm_rx_fifo_handler() implies that there are queued IN transfers
which will re-arm the DMA in USB peripheral once they complete.

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>

josuah reviewed (commented):
Some feedback about the implementation.

However, it might make more sense to complete the discussion on this thread first.

- https://github.com/zephyrproject-rtos/zephyr/pull/99213/files#r2517323683

ndreys commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -128,20 +128,19 @@ struct cdc_acm_uart_data {
-static struct net_buf *cdc_acm_buf_alloc(struct usbd_class_data *const c_data,
-					 const uint8_t ep)
+static struct net_buf *__cdc_acm_buf_alloc(struct net_buf_pool *pool,
+					   const uint8_t ep)
```
Sure, trivial to rename

ndreys committed:
usb: device_next: cdc_acm: Clear RX FIFO busy bit on failure

We need to clear RX FIFO busy bit in these falure paths in order to
have an accurate state of the system.

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
